File: InterviewPrep-Python/CrackingTheCodingInterview/Chapter1-ArraysAndStrings/Q6_StringCompression.py
'''
Created on Dec 29, 2016

@author: drawkward
'''
import unittest
def StringCompression(input):
    
    count = 0
    output=""
    for i in range(len(input)):
        count+=1
        if(i+1>=len(input) or input[i]!=input[i+1]):
            output += input[i] + str(count)
            count=0
            
    if len(output) >= len(input):
        return input
    else:
        return output
    
class Testcase(unittest.TestCase):
    data = [("aabbccc", "a2b2c3"),
            ("aabbcc", "aabbcc")
            ]
    
    def test_function(self):
        for [input, output] in self.data:
            actual = StringCompression(input)
            self.assertEqual(output, actual)
if __name__ == '__main__':
    unittest.main()
        

File: InterviewPrep-Java/src/CrackingTheCodingInterview/Chapter1_ArraysAndStrings/Q6_StringCompression.java
package CrackingTheCodingInterview.Chapter1_ArraysAndStrings;

public class Q6_StringCompression {
	public static String StringCompression(String input){
		
		int count=0;
		StringBuilder output=new StringBuilder();
		for(int i=0;i<input.length();++i){
			count++;
			
			if(i+1>=input.length() || input.charAt(i)!=input.charAt(i+1)){
				output.append(input.charAt(i));
				output.append(count);
				count=0;
			}
		}
		if(output.length()<input.length()) return output.toString();
		else return input;
	}
	
	public static void main(String args[]){
		System.out.print(StringCompression("aabbcc"));
	}
}
